### ISSUE TITLE
Error when files get staged

### ISSUE THREAD

**Issue by mcmillion** [lewis6991/gitsigns.nvim#571]
### Description

Getting the following error when files are staged (via fugitive):

```
Error executing vim.schedule lua callback: ...e/pack/packer/start/gitsigns.nvim/lua/gitsigns/async.lua:75: The coroutine failed with this message: ...pack/packer/start/gitsigns.nvim/lua/gitsigns/manager.lua:296: Fa
iled to rename buffer                                                                                                                                                                                                 
stack traceback:                                                                                                                                                                                                      
        [C]: in function 'nvim_buf_set_name'                                                                                                                                                                          
        ...pack/packer/start/gitsigns.nvim/lua/gitsigns/manager.lua:296: in function 'handle_moved'                                                                                                                   
        ...pack/packer/start/gitsigns.nvim/lua/gitsigns/manager.lua:340: in function <...pack/packer/start/gitsigns.nvim/lua/gitsigns/manager.lua:304>                                                                
stack traceback:                                                                                                                                                                                                      
        [C]: in function 'error'                                                                                                                                                                                      
        ...e/pack/packer/start/gitsigns.nvim/lua/gitsigns/async.lua:75: in function <...e/pack/packer/start/gitsigns.nvim/lua/gitsigns/async.lua:70>
```

### Neovim version

NVIM v0.7.0

### Operating system and version

macOS 12.4

### Expected behavior

No errors

### Actual behavior

Errors

### Minimal config

```Lua
vim.o.packpath = '/tmp/nvim/site'

local plugins = {
  gitsigns = 'https://github.com/lewis6991/gitsigns.nvim',
  -- ADD OTHER PLUGINS _NECESSARY_ TO REPRODUCE THE ISSUE
}

for name, url in pairs(plugins) do
  local install_path = '/tmp/nvim/site/pack/test/start/'..name
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
end

require('gitsigns').setup{
  debug_mode = true, -- You must add this to enable debug messages
  -- ADD GITSIGNS CONFIG THAT IS _NECESSARY_ FOR REPRODUCING THE ISSUE
}

-- ADD INIT.LUA SETTINGS THAT IS _NECESSARY_ FOR REPRODUCING THE ISSUE
```


### Steps to reproduce

Not really sure here, happens after some amount of work and then staging files. My assumption is some files have either been deleted or renamed and that triggers this

### Gitsigns debug messages

_No response_

**Comment by lewis6991** [comment]
This will be difficult to resolve without repro steps.

I've looked into the Neovim source and this error mostly only happens when a `BufFilePre` autocmd runs, either if it changes the current buffer, or another error occurs. Another case this error happens is when the filename we are changing to already exists.

**Comment by Yutsuten** [comment]
Recently I got this error frequently, I was refactoring several scripts, and every time I had deleted or moved files. It happens exatly when you press `s` to stage the file.

```
Error executing vim.schedule lua callback: ...nvim/site/pack/all/start/gitsigns/lua/gitsigns/async.lua:67: The coroutine failed with this message: ...im/site/pack/all/start/gitsigns/lua/gitsigns/manager.lua:297: Failed to rename buffer
stack traceback:
        [C]: in function 'nvim_buf_set_name'
        ...im/site/pack/all/start/gitsigns/lua/gitsigns/manager.lua:297: in function 'handle_moved'
        ...im/site/pack/all/start/gitsigns/lua/gitsigns/manager.lua:341: in function <...im/site/pack/all/start/gitsigns/lua/gitsigns/manager.lua:305>
stack traceback:
        [C]: in function 'error'
        ...nvim/site/pack/all/start/gitsigns/lua/gitsigns/async.lua:67: in function <...nvim/site/pack/all/start/gitsigns/lua/gitsigns/async.lua:62>
```

Its kind of annoying, so I'll try to find a way to reproduce it.

**Comment by Yutsuten** [comment]
Its tricky, but I somehow understood the conditions for this bug to happen.

1. The buffer of a file that was deleted or moved must be still open in nvim **AND** not updated (showing the old contents)
2. Stage/unstage a file using the `:G` window from [vim-fugitive](https://github.com/tpope/vim-fugitive)

Once you raise the error, you can repeat it endlessly by repeatedly unstaging and staging from `:G` window.

Steps:

```shell
git init
echo "some text" > file1
nvim file1  # Do nothing, just quit
<Ctrl + Z>
git add . && git commit -m "first commit"
mv file1 file2
fg
```

Then keep both `file1` and `file2` in different windows. It is important that `file1` is still opened and that the old contents are still being shown (do not `:e` on it).

Press `:G` and stage/unstage the files.

---

**Note:** When I raised this error in my refactoring, the deleted file's window wasn't really open, but vim keeps the buffer opened even if it is not being shown in a window, so... If programming for a long time and you removed some files that are still open in nvim, you _will_ see the error.

**Comment by lewis6991** [comment]
I just tried the above steps and was unable to reproduce.

**Comment by Yutsuten** [comment]
I recorded a GIF reproducing the bug.

https://i.imgur.com/mXi6rFW.gifv

**Comment by lewis6991** [comment]
Sorry, but the gif isn't very helpful.

Please provide a minimal config and **exact** steps as outlined by the issue template.

**Comment by Yutsuten** [comment]
OK here we go.

This is a detailed step-by-step with a minimal configuration.

### 1. Clean configuration

Cleaned all the nvim [configuration I have](https://github.com/Yutsuten/linux-config).

```shell
rm -f ~/.config/nvim/init.vim
rm -rf ~/.local/share/nvim/site/*
rm -rf ~/.config/nvim/*
```

### 2. Clone vim-fugitive and gitsigns.nvim

```shell
mkdir -p ~/.local/share/nvim/site/pack/all/start
cd ~/.local/share/nvim/site/pack/all/start
git clone https://github.com/tpope/vim-fugitive.git
git clone https://github.com/lewis6991/gitsigns.nvim.git
```

### 3. Add gitsigns.nvim config

```shell
mkdir ~/.local/share/nvim/site/plugin
nvim ~/.local/share/nvim/site/plugin/gitsigns.lua
```

`gitsigns.lua` content:

```lua
require('gitsigns').setup({})
```

Then go repeat the steps I described in https://github.com/lewis6991/gitsigns.nvim/issues/571#issuecomment-1147239899 and showed in the GIF.

**Comment by lewis6991** [comment]
That is not a minimal config, that is just *your* config.

> Please provide a minimal config and exact steps as outlined by the issue template.

Place your minimal config here by following the instructions in the comments.

```lua
vim.o.packpath = '/tmp/nvim/site'

local plugins = {
  gitsigns = 'https://github.com/lewis6991/gitsigns.nvim',
  -- ADD OTHER PLUGINS _NECESSARY_ TO REPRODUCE THE ISSUE
}

for name, url in pairs(plugins) do
  local install_path = '/tmp/nvim/site/pack/test/start/'..name
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
end

require('gitsigns').setup{
  debug_mode = true, -- You must add this to enable debug messages
  -- ADD GITSIGNS CONFIG THAT IS _NECESSARY_ FOR REPRODUCING THE ISSUE
}

-- ADD INIT.LUA SETTINGS THAT IS _NECESSARY_ FOR REPRODUCING THE ISSUE
```

And provide steps that involve the above config (edited with what is required) which can then be run with `nvim --clean -u minimal.lua`.

**Comment by Yutsuten** [comment]
I was about to send the configuration now, sorry for not noticing you wanted in that format.

You wrote it already and it is probably unnecessary, but I just reproduced the bug with this config:

```lua
vim.o.packpath = '/tmp/nvim/site'

local plugins = {
  gitsigns = 'https://github.com/lewis6991/gitsigns.nvim',
  vimfugitive = 'https://github.com/tpope/vim-fugitive'
}

for name, url in pairs(plugins) do
  local install_path = '/tmp/nvim/site/pack/test/start/'..name
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
end

require('gitsigns').setup{
  debug_mode = true, -- You must add this to enable debug messages
}
```

Thank you for taking your time to fix this bug. Hope you have a good day.

**Comment by lewis6991** [comment]
I've attempted to fix this in #672. @Yutsuten could you please test this?
